test/distributed/optim/test_zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,511 @@
+
+
+class TestZeroRedundancyOptimizerThreeRanks(TestZeroRedundancyOptimizer):
+    @property
+    def world_size(self):
+        return 3
+
+    def test_add_param_group(self):
+        self.dist_init(self.rank)
+
+        params = []
+        for size in [4, 5, 2, 6, 4]:
+            params.append(torch.rand(size, 1))
+        o = ZeroRedundancyOptimizer(params, optim=SGD, lr=0.1)
+        self.assertEqual(len(o.param_groups), 1)
+        o.add_param_group({"params": [torch.rand(3, 1)]})
+        self.assertEqual(len(o.param_groups), 2)
+
+        # Verify that added group is added to the correct partition making all have 8 elements.
+        self.assertEqual(sum([x.numel() for g in o.optim.param_groups for x in g["params"]]), 8)
+        self.assertEqual(len(o.optim.param_groups), 2)
+
+    def test_sharding(self):
+        self.dist_init(self.rank)
+        params = []
+        for size in [5, 4, 2, 6, 4, 3]:
+            params.append(torch.rand(size, 1))
+        o = ZeroRedundancyOptimizer(params, optim=SGD, lr=0.1)
+        self.assertEqual(sum([x.numel() for x in o.optim.param_groups[0]["params"]]), 8)
+
+    def test_collect_shards(self):
+        self.dist_init(self.rank)
+        device = torch.device(self.rank) if torch.cuda.device_count() > 1 else torch.device("cpu")
+        reference_rank = 0
+
+        # Run a dummy step so that the optimizer state dict exists
+        batch, input_width, hidden, target_width = 3, 20, 10, 5
+        target = torch.rand((batch, target_width), device=device)
+        inputs = torch.rand((batch, input_width), device=device)
+
+        model = torch.nn.Sequential(torch.nn.Linear(input_width, hidden), torch.nn.Linear(hidden, target_width))
+        model.to(device)
+
+        loss_fn = torch.nn.L1Loss()
+        loss_fn.to(device)
+
+        # With SGD, Momentum is required to get a state to shard
+        optimizer = ZeroRedundancyOptimizer(model.parameters(), optim=SGD, lr=0.1, momentum=0.99)
+
+        def closure():
+            optimizer.zero_grad()
+            output = model(inputs)
+            loss = loss_fn(output, target)
+            loss.backward()
+            return loss
+
+        _ = optimizer.step(closure=closure)
+
+        # Update the optimizer state on the reference rank
+        optimizer.consolidate_state_dict(recipient_rank=reference_rank)
+
+        # Fetch the state on the reference rank
+        # - check that it has the correct size
+        # - load it again
+        if self.rank == reference_rank:
+            optimizer_state_dict = optimizer.state_dict()
+            self.assertEqual(len(optimizer_state_dict["state"]), self.world_size)
+        else:
+            optimizer_state_dict = {}
+
+        optimizer_state_dict = _broadcast_object(
+            optimizer_state_dict, src_rank=reference_rank, group=dist.group.WORLD, dist_device=device
+        )
+
+        # Load the optimizer state dict
+        optimizer.load_state_dict(optimizer_state_dict)
```

Comment:
Shouldn't we assert something after this?

Reply:
I'll change to something better, idea from @joshim5 : something like checkpoint after 2 iterations, iterate once or twice, save state then load from checkpoint and re-iterate, check that we get to the same state. Would that look ok ? Checks save and load e2e